skbio/metadata/_mixin.py

```diff
@@ -406,3 +407,114 @@ def has_positional_metadata(self):
         """
         return (self._positional_metadata is not None and
                 len(self.positional_metadata.columns) > 0)
+
+
+class IntervalMetadataMixin(metaclass=abc.ABCMeta):
+    def _interval_metadata_axis_len_(self):
+        ''''''
+        raise NotImplementedError
+
+    @abc.abstractmethod
+    def __init__(self, interval_metadata=None):
+        raise NotImplementedError
+
+    def _init_(self, interval_metadata=None):
+        if interval_metadata is None:
+            # Could use deleter but this is less overhead and needs to be fast.
+            self._interval_metadata = None
+        else:
+            # Use setter for validation and copy.
+            self.interval_metadata = interval_metadata
+
+    @property
+    @experimental(as_of="0.5.0-dev")
+    def interval_metadata(self):
+        '''``IntervalMetadata`` object containing info about interval features.
+        '''
+        if self._interval_metadata is None:
+            # Not using setter to avoid copy.
+            self._interval_metadata = IntervalMetadata(
+                self._interval_metadata_axis_len_())
```

Comment:
My comment here no longer makes sense given the interpretation of exclusive upper bounds. It's great that an empty object (e.g. Sequence) can have empty positional and interval metadata!

I don't think we should allow start/end bounds that are the same (e.g. `(0, 0)`) because it doesn't represent any place in the coordinate space and users won't be able to query one of these intervals with `.query(locations=[(0,0)]`. Seems clearer to disallow these types of intervals entirely. It also matches how positional metadata works -- you can't have metadata attached to an "empty" position.